### DISCUSSION TITLE
Bind group layout gives weird results

### DISCUSSION THREAD

**Discussion by Rus-Abd** [gpuweb/gpuweb#4619] [Q&A] [answered]
When I pass write my mouse coords as floats, everything works correct
```rs
struct Uniforms { resolution: vec2f, time: f32, x:f32,y:f32 };
@binding(0) @group(0) var<uniform> U : Uniforms;
  
@fragment
fn main(@builtin(position) FragCoord : vec4f) -> @location(0) vec4<f32> {
      var uv = FragCoord.xy/U.resolution * 2. -1.;
      uv*= vec2f(1, -1);
      var x = U.x;
  return vec4<f32>(x,x,x,1.);
}

```
But when I use them as vec2f, my mouse X coordinate gets written into the Y of my mouse uniform
```rs
struct Uniforms { resolution: vec2f, time: f32, mouse: vec2f };
@binding(0) @group(0) var<uniform> U : Uniforms;
  
@fragment
fn main(@builtin(position) FragCoord : vec4f) -> @location(0) vec4<f32> {
      var uv = FragCoord.xy/U.resolution * 2. -1.;
      uv*= vec2f(1, -1);
      var x = U.mouse.x;
  return vec4<f32>(x,x,x,1.);
}
```
Any ideas why this is happening?
And Additional question, I'm getting an error if I specify the size 5*4=20 for my buffer, and it requires 24 size, why?
```js
    this.uniformsBuffer = this.device.createBuffer({
      size: 6 * 4,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
    });

    this.uniformsBindGroup = this.device.createBindGroup({
      layout: this.pipeline.getBindGroupLayout(0),
      entries: [
        {
          binding: 0,
          resource: { buffer: this.uniformsBuffer },
        },
      ],
    });

    this.uniformsArrayBuffer = new Float32Array(5);
    this.uniformsArrayBuffer[0] = this._canvas.width;
    this.uniformsArrayBuffer[1] = this._canvas.height;
    this.uniformsArrayBuffer[2] = this.time;
    this.uniformsArrayBuffer[3] = this.mouse.x;
    this.uniformsArrayBuffer[4] = this.mouse.y;
```

**Comment by ben-clayton** [comment, marked as answer]
[`vec2f` has different alignment requirements to `f32`](https://www.w3.org/TR/WGSL/#alignment-and-size)

`f32` has a 4-byte size and alignment.
`vec2f` has a 8-byte size, and 8-byte alignment.

```ts
struct Uniforms { 
/* [0 .. 7]   */ resolution: vec2f, 
/* [8 .. 12]  */ time: f32,
/* [12 .. 16] */ x: f32,
/* [16 .. 20] */ y: f32,
}
```

```ts
struct Uniforms { 
/* [0 .. 7]   */ resolution: vec2f, 
/* [8 .. 12]  */ time: f32,
/* [12 .. 16] */ // implicit padding due to vec2f's alignment
/* [16 .. 23] */ mouse: vec2f,
}
```

**Reply by Rus-Abd** [reply]
Thanks!
Thought that everything aligned to 4-byte, just used different sizes.

**Reply by greggman** [reply]
fyi, you can use the [WGSL Offset Computer](https://webgpufundamentals.org/webgpu/lessons/resources/wgsl-offset-computer.html#x=5d00000100ec00000000000000003d888b0237284d3025f2381bcb288ae878c60ccc524b949619433c114b6202c5a357134497c84203e3f26302a85e901aab011ed2d91ee7c48a2e561c53cfe50586dcd8666cf0d31023c0c03081b18e013de9168e3848e7b57bc1be57b454d53add6f11c11ea1525f348ed66d08687c72f9c1df1c396e19824845aa8bc221f21498e748571705a93b03f1bc9bfe95cc2ed1aeda5d25e725613b40dd80ef7c3083ff03bfff2c86fc00) to check alignments
